include/seqan3/io/sam_file/format_bam.hpp

```diff
@@ -961,6 +945,56 @@ inline void format_bam::write_alignment_record([[maybe_unused]] stream_type & st
     } // if constexpr (!detail::decays_to_ignore_v<header_type>)
 }
 
+//!\copydoc seqan3::detail::format_sam_base::write_header
+template <typename stream_t, typename header_type>
+inline void format_bam::write_header(stream_t & stream, sam_file_output_options const & options, header_type & header)
+{
+    if constexpr (detail::decays_to_ignore_v<header_type>)
+    {
+        throw format_error{"BAM can only be written with a header but you did not provide enough information! "
+                           "You can either construct the output file with reference names and reference length "
+                           "information and the header will be created for you, or you can access the `header` member "
+                           "directly."};
+    }
+    else
+    {
+        detail::fast_ostreambuf_iterator stream_it{*stream.rdbuf()};
+
+        std::ranges::copy_n("BAM\1", 4, stream_it);
+
+        // write SAM header to temporary stream first to query its size.
+        std::ostringstream os;
+        detail::format_sam_base::write_header(os, options, header);
+#if defined(__GNUC__) && (__GNUC__ == 10)
+        int32_t const l_text{static_cast<int32_t>(os.str().size())};
+#else
+        int32_t const l_text{static_cast<int32_t>(os.view().size())};
+#endif
+        std::ranges::copy_n(reinterpret_cast<char const *>(&l_text), 4, stream_it); // write text length
+
+#if defined(__GNUC__) && (__GNUC__ == 10)
+        auto header_view = os.str();
+#else
+        auto header_view = os.view();
+#endif
+        std::ranges::copy(header_view, stream_it);
+
+        int32_t const n_ref{static_cast<int32_t>(header.ref_ids().size())};
```

Comment:
Do we do bounds checking somewhere? For the number of references and the length of the references?

Reply:
I added asserts